[July 27-8, 2021] CodeQL package manager: update CodeQL CLI docs (beta) (#6339)
github/codeql: docs/codeql/codeql-cli/analyzing-databases-with-the-codeql-cli.rst

Before starting an analysis you must:

- :doc:`Set up the CodeQL CLI <getting-started-with-the-codeql-cli>` so that it can find the queries
  and libraries included in the CodeQL repository.
- :doc:`Create a CodeQL database <creating-codeql-databases>` for the source
  code you want to analyze.

> Running ``codeql database analyze``

Reviewer: Ah..I see you have it down below.

you could use the following command from the directory containing your database::

   codeql database analyze <javascript-database> ../ql/javascript/ql/src/Declarations/UnusedVariable.ql --format=csv --output=js-analysis/js-results.csv

This command runs a simple query that finds potential bugs related to unused
variables, imports, functions, or classes---it is one of the JavaScript
queries included in the CodeQL repository. You could run more than one query by
specifying a space-separated list of similar paths.

The analysis generates a CSV file (``js-results.csv``) in a new directory
(``js-analysis``).

You can also run your own custom queries with the ``database analyze`` command.
For more information about preparing your queries to use with the CodeQL CLI,
see ":doc:`Using custom queries with the CodeQL CLI <using-custom-queries-with-the-codeql-cli>`."

Running a CodeQL pack
~~~~~~~~~~~~~~~~~~~~~

.. include:: ../reusables/beta-note-package-management.rst

To run an existing CodeQL query pack from the GitHub Container registry, you need to download it first::

   codeql pack download microsoft/coding-standards@1.0.0

Afterwards, you can run the pack on a specific database::

>    codeql database analyze <database> microsoft/coding-standards@1.0.0 <scope>/<other-pack> --format=sarifv2.1.0 --output=query-results.sarif

Reviewer 2: @aeisenberg, I am curious about this example. I went to https://github.com/orgs/microsoft/packages?tab=packages&q= and can not see this pack there. Are they going to publish it later?

Reviewer 2: May I suggest to use something like `codeql database analyze <database> codeql/csharp-queries <scope>/<other-pack> --format=sarifv2.1.0 --output=query-results.sarif` example instead?

To run a single query over a CodeQL database for a JavaScript codebase,
you could use the following command from the directory containing your database::

   codeql database analyze <javascript-database> ../ql/javascript/ql/src/Declarations/UnusedVariable.ql --format=csv --output=js-analysis/js-results.csv

This command runs a simple query that finds potential bugs related to unused
variables, imports, functions, or classes---it is one of the JavaScript
queries included in the CodeQL repository. You could run more than one query by
specifying a space-separated list of similar paths.

The analysis generates a CSV file (``js-results.csv``) in a new directory
(``js-analysis``).

You can also run your own custom queries with the ``database analyze`` command.
For more information about preparing your queries to use with the CodeQL CLI,
see ":doc:`Using custom queries with the CodeQL CLI <using-custom-queries-with-the-codeql-cli>`."

> Running a CodeQL pack

Reviewer 3: Resolving this comment in favor of https://github.com/github/codeql/pull/6339/files/87005b93e3511ae4fcfb654140fe90ec03ecf3ef#r675215208.

~~~~~~~~~~~~~~~~~~~~~

.. pull-quote::

    Note

>     The CodeQL package manager is currently in beta and subject to change. During the beta, CodeQL packs are available only in the GitHub Package Registry (GHPR). You must use version 2.5.8 or later of the CodeQL CLI to use the CodeQL package manager.

Reviewer 3: Ah, of course. This has always been confusing to me: GitHub Packages is the GitHub Container Registry (GHCR). 😃

    During the beta, you must specify ``--allow-packs`` whenever you run the ``analyze`` command.

To run an existing CodeQl pack from the GitHub Package Registry (GHPR), you need to download it first:

   codeql pack download microsoft/coding-standards@1.0.0

Afterwards, you can run the pack on a specific database:

>    codeql database analyze <database> --allow-packs microsoft/coding-standards@1.0.0 <scope>/<other-pack>

Reviewer 3: Good catch! ✨

The ``analyze`` command above runs the default suite from ``microsoft/coding-standards v1.0.0`` and the latest version of ``scope/other-pack`` on the specified database.
For further information about default suites, see ":ref:`Publishing and using CodeQL packs <publishing-and-using-codeql-packs>`".

> Running GitHub code scanning suites

Reviewer: I think these should be called _query suites_.

And again, separating out running queries, packs, and query suites into 3 sections seems artificial, since you can combine all three of them in a single cli invocation. I'd rather see a single section that includes a description of the different ways of running `database analyze`.

And then you can have a separate section (if necessary) that describes `github upload-results`

Reviewer: To be clear, my suggestion above might be out of scope for this PR. If you don't have time to address this before the packaging release, then it's fine to think about it for later.

Reviewer 3: This is a great suggestion. I was wondering how to best structure or group the different ways to analyze databases. I'm going to focus on fixing this PR from a functional perspective first and note this as a work item for the next iteration. ✨

Reviewer 3: Once the diff is outdated, this comment thread might be lost, so I moved it over to the main thread:
https://github.com/github/codeql/pull/6339#issuecomment-886510387

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To run the GitHub code scanning suite of queries over a CodeQL database for a C/C++ codebase,
you could use the following command from the directory containing your database::

   codeql database analyze <cpp-database> cpp-code-scanning.qls --format=sarifv2.1.0 --output=cpp-results.sarif

The analysis generates a file in the v2.1.0 SARIF format that is supported by all versions of GitHub.

> This file can be uploaded to GitHub using ``github upload-results`` or the code scanning API.

Reviewer: ```suggestion
This file can be uploaded to GitHub executing ``codeql github upload-results`` or the code scanning API.
```

    Note

>     The CodeQL package manager is currently in beta and subject to change. During the beta, CodeQL packs are available only using GitHub Packages - the GitHub Container registry. You must use version 2.5.8 or later of the CodeQL CLI to use the CodeQL package manager.

Reviewer: I _think_ we want to specify the 2.6.0 pre-release. There are several other places to update as well.

   ql/src/codeql-suites/go-code-scanning.qls

The repository also includes the query suites used by `LGTM.com <https://lgtm.com>`__.

> These are stored alongside the query suites with names of the form: ``<language>-lgtm.qls``.

Reviewer 4: ```suggestion
These are stored alongside the query suites for code scanning with names of the form: ``<language>-lgtm.qls``.
```

If the analysis found fewer results for standard queries than you expected, review the results of the diagnostic and summary queries to check whether the CodeQL database is likely to be a good representation of the codebase that you want to analyze.

> Integrating a CodeQL pack into a code scanning workflow in GitHub
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 4: Fix tests.

```suggestion
Integrating a CodeQL pack into a code scanning workflow in GitHub
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

.. pull-quote::

    Note

    The CodeQL package manager is currently in beta and subject to change. During the beta, CodeQL packs are available only in GitHub Packages - the GitHub Container Registry (GHCR). You must use version 2.5.8 or later of the CodeQL CLI to use the CodeQL package manager.

Using packs inside your Code Scanning setup allows selecting query packs from various sources.

In the CodeQL configuration file, the section called ``packs`` holds a list of CodeQL package references. Each package listed there will be downloaded by the action and the its default suite will be run.

   queries:
     - queries/query1.ql
     - queries/suite1.qls
   packs:
     - codeql/pack1@~1.2.3  # latest version compatible with 1.2.3
     - codeql/pack2         # latest version

For multi-language runs, you can specify a nested map of packs:

   packs:
     javascript:
       - codeql/js-pack1@~1.2.3
       - codeql/js-pack2
     java:
       - codeql/java-pack1@~1.2.3
       - codeql/java-pack2

In the CodeQL workflow file, you can add ``packs`` as input for the ``init`` action by using a comma-separated list of CodeQL packages and optional versions.

>    uses: github/codeql-action@v1
>    with:
>      packs: codeql/pack1@~1.2.3, +codeql/pack2

Reviewer 4: To fix the first Sphinx generation error:

```suggestion
.. code-block:: none

   uses: github/codeql-action@v1
   with:
     packs: codeql/pack1@~1.2.3, +codeql/pack2
```

If the analysis found fewer results for standard queries than you expected, review the results of the diagnostic and summary queries to check whether the CodeQL database is likely to be a good representation of the codebase that you want to analyze.

Integrating a CodeQL pack into a Code Scanning workflow
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. pull-quote::

    Note

    The CodeQL package manager is currently in beta and subject to change. During the beta, CodeQL packs are available only in the GitHub Package Registry (GHPR). You must use version 2.5.8 or later of the CodeQL CLI to use the CodeQL package manager.

> Using packs inside your Code Scanning setup allows selecting query packs from various sources.

Reviewer: Are we calling them "CodeQL packs"?

Reviewer 3: Currently, the docs are inconsistent in this regard and a decision hasn't been made.
Personally, I like the proposal to go with _"CodeQL packages"_ which include _"QL packs"_.

Reviewer 3: In order to not lose this once the diff changes, I added the following comment to the main PR thread:
https://github.com/github/codeql/pull/6339#issuecomment-886513565

In the CodeQL configuration file, the section called ``packs`` holds a list of CodeQL package references. Each package listed there will be downloaded by the action and the its default suite will be run.

   queries:
     - queries/query1.ql
     - queries/suite1.qls
   packs:
     - codeql/pack1@~1.2.3  # latest version compatible with 1.2.3
     - codeql/pack2         # latest version

For multi-language runs, you can specify a nested map of packs:

   packs:
     javascript:
       - codeql/js-pack1@~1.2.3
       - codeql/js-pack2
     java:
       - codeql/java-pack1@~1.2.3
       - codeql/java-pack2

In the CodeQL workflow file, you can add ``packs`` as input for the ``init`` action by using a comma-separated list of CodeQL packages and optional versions.

   uses: github/codeql-action@v1
   with:

>      packs: codeql/pack1@~1.2.3, codeql/pack2

Reviewer: Note that you can prefix with a `+`. It has the same meaning as when you prefix the `queries` input with a `+`. 

With no `+`, the input overwrites the config file. With a `+`, the input is _combined_ with the config file.

> This format does not yet support multi-language analyses.

Reviewer: ```suggestion
This format does not support multi-language analyses.
```

(probably, it never will)

> Running all queries in a directory

Reviewer: Again, can we combine this?

Reviewer 3: See https://github.com/github/codeql/pull/6339/files/87005b93e3511ae4fcfb654140fe90ec03ecf3ef#r675215208
